x/poolmanager/taker_fee.go

```diff
@@ -176,3 +176,110 @@ func CalcTakerFeeExactOut(tokenIn sdk.Coin, takerFee osmomath.Dec) (sdk.Coin, sd
+// The function returns an error if the total taker fee share percentage is greater than 1.
+//
+// Parameters:
+// - ctx: The context of the function call.
+// - denomsInvolvedInRoute: A slice of strings representing the denominations involved in the route.
+// - totalTakerFees: The total taker fees from the swap represented as sdk.Coins.
+//
+// Returns:
+// - An error if the total taker fee share percentage is greater than 1, or if there's an error in increasing the accumulator for the denomination pair.
+func (k Keeper) TakerFeeSkim(ctx sdk.Context, denomsInvolvedInRoute []string, totalTakerFees sdk.Coins) error {
+	// Sort the denoms involved in the route lexicographically.
+	osmoutils.SortSlice(denomsInvolvedInRoute)
+
+	// Retrieve the share agreements for denoms.
+	denomShareAgreements, alloyedAssetShareAgreements := k.getTakerFeeShareAgreements(denomsInvolvedInRoute)
+
+	// Process denom share agreements if any exist.
+	if len(denomShareAgreements) > 0 {
+		return k.processDenomShareAgreements(ctx, denomShareAgreements, totalTakerFees)
+	}
+
+	// Process alloyed asset share agreements if no denom share agreements exist.
+	if len(alloyedAssetShareAgreements) > 0 {
+		return k.processAlloyedAssetShareAgreements(ctx, alloyedAssetShareAgreements, totalTakerFees)
+	}
+
+	return nil
+}
+
+// getTakerFeeShareAgreements checks for individual denomShareAgreement and alloyedAssetShareAgreement denoms.
+func (k Keeper) getTakerFeeShareAgreements(denomsInvolvedInRoute []string) ([]types.TakerFeeShareAgreement, []types.TakerFeeShareAgreement) {
+	denomShareAgreements := []types.TakerFeeShareAgreement{}
+	alloyedAssetShareAgreements := []types.TakerFeeShareAgreement{}
+
+	for _, denom := range denomsInvolvedInRoute {
+		// We first check if this denom has a taker fee share agreement.
+		takerFeeShareAgreement, found := k.getTakerFeeShareAgreementFromDenom(denom)
+		if found {
+			// If the denom has a denomShareAgreement, add the denomShareAgreement to the denomShareAgreements slice.
+			denomShareAgreements = append(denomShareAgreements, takerFeeShareAgreement)
+		} else if len(denomShareAgreements) == 0 {
+			// If there are no denomShareAgreements in the denomShareAgreements slice, continue to filter this denom to determine if it is a registered alloyedAssetShareAgreement denom.
+			// If there are 1 or more denomShareAgreements in the denomShareAgreements slice, we don't need to check for alloyedAssetShareAgreements anymore, since the taker fee share
+			// only goes to alloyedAssetShareAgreements IF there are no denomShareAgreement denoms in the route.
+
+			// Check if the denom is an alloyedAssetShareAgreement denom.
+			// If it is, add the alloyedAssetShareAgreement to the alloyedAssetShareAgreements slice.
+			cachedAlloyContractState, found := k.getRegisteredAlloyedPoolFromDenom(denom)
+			if found {
+				alloyedAssetShareAgreements = append(alloyedAssetShareAgreements, cachedAlloyContractState.TakerFeeShareAgreements...)
+			}
+		}
+	}
```

Comment:
Based on this logic, if regular denom appears before alloyed denom and both have shared agreements, the alloyed share agreement will not be returned. However, if alloyed denom appears first in `denomsInvolvedInRoute ` , both will be returned.

I don't understand why that is the case. Could you please explain and document this?

Reply:
I think that we might want to first iterate and the `denomsInvolvedInRoute` in-full and only then check for the alloyed.

Otherwise, we might have the inconsistency in logic from my previous comment.

Reply:
Based on my understanding , I would expect [this](https://github.com/osmosis-labs/osmosis/commit/551ea6a4ace24f9021e8a407406c40626753cb1c) test case and [this one](https://github.com/osmosis-labs/osmosis/blob/551ea6a4ace24f9021e8a407406c40626753cb1c/x/poolmanager/taker_fee_test.go#L294-L303) to produce the same results.

However, that is not the case. Could you please let me know what is the intended behavior here?

Reply:
I think it only works as intended due to the context in which its called. See https://github.com/osmosis-labs/osmosis/blob/d449a8a7613f74360301f2571702aab4823131fc/x/poolmanager/taker_fee.go#L201-L20. I think its done strictly for performance reasons, since we only need alloyedAssetShareAgreements IFF there are no denomShareAgreements (I think of denomShareAgreements as tier1 and alloyedAssetShareAgreements as tier2, and we only involve tier2 agreements in the event that tier1 does not exist).

You are right to point out though that the caller doesn't really know this logic is going on. Are you suggesting we don't filter and just return all denomShareAgreements and alloyedAssetShareAgreements, even if the alloyedAssetShareAgreements wont be used? Or do you think we should keep the logic as is, and maybe at the end of the method in getTakerFeeShareAgreements just return a blank alloyedAssetShareAgreements if there are denomShareAgreements?

Finally, I am a little confused on your comment about the two tests. You said "I would expect these two test cases to produce the same results", but at least from what I see they do produce the same results. Maybe I am misunderstanding?

Reply:
From our call, this commit should resolve this issue, thanks for the discussion: https://github.com/osmosis-labs/osmosis/pull/8310/commits/4cbec3d93c6db659b2f167afc4a7dfff8c06a9a6